src/index.js

```diff
@@ -0,0 +1,360 @@
+ *  - https://www.mongodb.com/blog/post/pitfalls-and-workarounds-for-tailing-the-oplog-on-a-mongodb-sharded-cluster
+ */
+
+const MongoDB = require('mongodb');
+const MongoOplog = require('mongo-oplog');
+const EventEmitter = require('eventemitter3');
+const Promise = require('bluebird');
+const RedisLock = require('redislock');
+const Debug = require('debug');
+
+const Timestamp = MongoDB.Timestamp;
+const debug = Debug('MongoOplogReader');
+
+const defaults = {
+  keyPrefix: 'mongoOplogReader',
+  ttl: 60 * 60 * 24, // amount of time in seconds to keep track of the emit status of an op
+  redundancy: 1,
+  masterDuration: 30, // in seconds
+  healthcheckDuration: 10 // in seconds
+};
+
+const defaultMongoConnectOptions = {
+  server: {
+    poolSize: 100,
+    auto_reconnect: true,
+    socketOptions: {
+      keepAlive: 5000,
+      connectTimeoutMS: 30000,
+      socketTimeout: 30000
+    }
+  }
+};
+
+class MongoOplogReader extends EventEmitter {
+
+  /**
+   * Create a new MongoOplogReader
+   * @param options
+   *  - connectionStrings {String[]} a list of mongodb connection strings
+   *  - redisClient {Object} an instance of the 'redis' module
+   *  - redundancy {Integer} number of processes per shard
+   *  - ttl {Integer} max time it may take a cluster to recover (in seconds)
+   *  - keyPrefix {String} the redis key prefix (default: 'mongoOplogReader')
+   */
+  constructor(options) {
+    super();
+    if (!options.redisClient || !options.redisClient.sadd) {
+      throw new Error('Invalid redisClient.');
+    }
+    this.connectionStrings = options.connectionStrings;
+    this.redisClient = Promise.promisifyAll(options.redisClient);
+    this.lock = RedisLock.createLock(options.redisClient);
+    this.oplogs = {};
+    this.replSets = {};
+    this.redundancy = options.redundancy || defaults.redundancy;
+    this.keyPrefix = options.keyPrefix || defaults.keyPrefix;
+    this.ttl = options.ttl || defaults.ttl;
+    this.masterDuration = options.masterDuration || defaults.masterDuration;
+    this.healthcheckDuration = options.healthcheckDuration || defaults.healthcheckDuration;
+    this.mongoConnectOptions = options.mongoConnectOptions || defaultMongoConnectOptions;
+    this.workerId = `${Math.random()}`.substr(2);
+    this.workerRegistrationKey = `${this.keyPrefix}:workerIds`;
+    this.assignmentsByConnStr = {};
+    this.assignmentsByWorkerId = {};
+
+    if (!Number.isInteger(this.redundancy) || this.redundancy > 10) {
+      throw new Error(`Specified redundancy '${this.redundancy}' should be an integer less than 10.`);
+    }
+  }
+
+  start() {
+    if (!Array.isArray(this.connectionStrings) || !this.connectionStrings.length) {
+      throw new Error('There are no connectionStrings.');
+    }
+    return Promise.resolve()
+      .then(() => this.registerWorker())
+      .then(() => this.acquireMasterLock())
+      .then(() => this.startPolling());
+  }
+
+  startPolling() {
```

Comment:
Should we expose a stopPolling method that clears out `this.masterInterval`, `this.healthcheckInterval` and `setTimeout(() => this.acquireMasterLock(), randomDelay);`?

Reply:
You would never stop polling while continuing to tail the oplog. So, we could have a stop() method that stops everything, but I can't think of a reason for it. So I'd rather wait until we need it before adding it because it unnecessarily increases the surface area for bugs. Currently if you want to stop, you just stop the process.